### PR TITLE
Add Milestone Notifications

### LINKED ISSUE #1647: Milestone notifications
* Your post just reached 100 reactions!
* Your post just reached 1,000 views!
* Etc.

As a user I'd like to be notified at certain milestones. In terms of when to notify, I think it would be fun to do it in binary escalation.

Maybe start notifying people at 64 reactions and then 128, 256, 512, 1,024 etc.

For view counts, maybe start at 1,024 and work up from there.

If the odd numbers are too weird, I still think doubling makes sense in terms of frequency. 1000, 2000, 4000, 8000, 16000 etc.

This could be implemented as a script that gets run after reactions and/or page view update etc. and awards a milestone if the relevant one had rewards the relevant notification.

This would be a complement to badges, which you can receive once for certain achievements. But this would exist only as a notification.

### PR DESCRIPTION
**Pull request by Zhao-Andy**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This PR adds milestone notifications! Milestones are achieved when your post reaches the following:
- 1024, 2048, 4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, or 1048576 total views
- 64, 128, 256, 512, 1024, 2048, 4096, or 8192  total reactions

This feature accounts for users and organization posts, and is flexible for any other sort of milestones we want to add.

## Related Tickets & Documents
Resolves #1647 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
Screenshot of reaction and view milestone notifications

## [optional] What gif best describes this PR or how it makes you feel?
It's the classic Oprah "You get a car! You get a car! YOU GET A CAR!" gif

### CODE REVIEW

File: app/models/notification.rb [hunk 1]
```diff
@@ -252,6 +252,29 @@ def send_tag_adjustment_notification(tag_adjustment)
     end
     handle_asynchronously :send_tag_adjustment_notification
 
+    def send_milestone_notification(milestone_hash)
```
Comment (lightalloy) [inline app/models/notification.rb]: I would suggest moving the send notification code to a service object not to overload the model.

File: app/models/notification.rb [hunk 2]
```diff
@@ -252,6 +252,29 @@ def send_tag_adjustment_notification(tag_adjustment)
...
+      milestone_hash[:next_milestone] = next_milestone(milestone_hash)
+      return unless should_send_milestone?(milestone_hash)
+
+      Notification.create!(
+        user_id: milestone_hash[:article].user_id,
+        notifiable_id: milestone_hash[:article].id,
+        notifiable_type: "Article",
+        json_data: { article: article_data(milestone_hash[:article]) },
+        action: "Milestone::#{milestone_hash[:type]}::#{milestone_hash[:next_milestone]}",
```
[the remark this answers is not in the document]
Comment (Zhao-Andy) [reply app/models/notification.rb]: Thanks :)

File: app/models/notification.rb (hunk 1 above)
Comment (Zhao-Andy) [reply app/models/notification.rb]: Yeah, I agree. I think I'll move all of that code in a different PR.

**Comment by Zhao-Andy** [comment]
Added 6 random gifs from Giphy. They load it directly from Giphy, and link back to Giphy in new tab. 

## Preview:
screen shot 2019-02-28 at 1 43 56 pm

## 6 random gifs:

File: app/views/notifications/_milestone.html.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<% json_data = notification.json_data %>
+<div class="content notification-content badge-content">
+  <div class="badge-title">
+    Your post <a href="<%= json_data["article"]["path"] %>"><%= json_data["article"]["title"] %></a> just passed <%= notification.milestone_count %> <%= notification.milestone_type.pluralize.downcase %>!
+  </div>
+  <p class="badge-description milestone-emojis">🎉🎉🎉🎉🎉🎉🎉🎉</p>
+  <a href="https://media.giphy.com/media/<%= json_data["gif_id"] %>/giphy-downsized.gif" rel="noopener noreferrer" target="_blank">
```
Comment (benhalpern) [inline app/views/notifications/_milestone.html.erb]: The `href` here should giphy _page_ as opposed to the file itelf. That would be `https://giphy.com/gifs/<%= json_data["gif_id"] %>` as opposed to the `https://media.giphy.com/media...` link for the image.

Comment (Zhao-Andy) [reply app/views/notifications/_milestone.html.erb]:
```suggestion
  <a href="https:/giphy.com/gifs/<%= json_data["gif_id"] %>" rel="noopener noreferrer" target="_blank">
```

Example: https://giphy.com/gifs/y8Mz1yj13s3kI

I think it technically redirects to the full link: `https://giphy.com/gifs/excited-oprah-shouting-y8Mz1yj13s3kI`, which should be okay. Would this need a `data-no-instant` attribute?

Comment (benhalpern) [reply app/views/notifications/_milestone.html.erb]: > Would this need a `data-no-instant` attribute?

Nope Instaclick knows how to deal with external links.

File: app/assets/stylesheets/notifications.scss [hunk 4]
```diff
@@ -67,11 +67,23 @@
...
+                padding-top: 10px;
+                padding-bottom: 0px;
+              }
             }
             img.badge-image{
               margin:10px 0px 20px;
               height: 150px;
             }
+            img.milestone-gif{
+                margin: 10px 0px 10px;
```
Comment (benhalpern) [inline app/assets/stylesheets/notifications.scss]: I think a border radius on the gif would be a nice touch.

```suggestion
                margin: 10px 0px 10px;
                border-radius: 8px;
```

**Comment by Zhao-Andy** [comment]
`final_final_final_pull_request_example.png`
final_final_final_pull_request_example

Border radius was a great idea 👍

**Comment by BogDAAAMN** [comment]
Finally!!!

**Pull request merged**